docs/variant.md

```diff
@@ -1,207 +1,200 @@
 
-Most data structures in most languages are about "this **and** that". A variant allows us to express "this **or** that".
+Variant types model values that may assume one of many known variations. This
+feature is similar to "enums" in other languages, but each variant form may
+optionally specify data that is carried along with it.
 
-```reason
-type myResponseVariant =
-  | Yes
-  | No
-  | PrettyMuch;
-
-let areYouCrushingIt = Yes;
-```
+This is a very powerful feature of the Reason language when used with
+[pattern matching](pattern-matching.md).
 
-`Yes`, `No` and `PrettyMuch` aren't strings, nor references, nor some special data type. They're called "constructors" (or "tag"). The `|` bar separates each constructor.
-
-**Note**: a variant's constructors need to be capitalized.
-
-## Usage
+## Creation
 
-Along with a variant comes one of the most important features of Reason, the `switch` expression.
+Variants require a type definition that specifies all of the "constructors",
+or "tags", for that variant. Constructors must start with an uppercase letter.
 
-A Reason `switch` is visually similar to other languages' `switch` (aka a large `if/elseif/elseif...`). It allows you to check every possible case of a variant. To use it, enumerate every variant constructor of the particular variant you'd like to use, each followed by an `=>` and the expression corresponding to that case.
+Here the constructors for the `animal` variant are `Cat`, `Dog`, `Horse`,
+and `Snake`:
 
 ```reason
-let message =
-  switch (areYouCrushingIt) {
-  | No => "No worries. Keep going!"
-  | Yes => "Great!"
-  | PrettyMuch => "Nice!"
-  };
-/* message is "Great!" */
+type animal =
+  | Cat
+  | Dog
+  | Horse
+  | Snake;
 ```
 
-A variant has an extremely rich amount of type system assistance. For example, we'll give you a type error if you've forgotten to cover a case of your variant, or if two cases are redundant. Be sure to check out switch and pattern-matching in a [later section](pattern-matching.md)!
-
-### Variant Needs an Explicit Definition
-
-If the variant you're using is in a different file, bring it into scope like you'd do [for a record](record.md#record-needs-an-explicit-definition):
+To create a value of type `animal` use one of the constructors:
 
 ```reason
-/* Zoo.re */
-
-type animal = Dog | Cat | Bird;
+let fluffy = Cat;
+let spot = Dog;
 ```
 
-```reason
-/* example.re */
+## Constructor Arguments
 
-let pet: Zoo.animal = Dog; /* preferred */
-/* or */
-let pet = Zoo.Dog;
-```
-
-### Constructor Arguments
-
-A variant's constructors can hold extra data separated by comma.
+Variant constructors can have data associated with them:
 
 ```reason
-type account =
-  | None
-  | Instagram(string)
-  | Facebook(string, int);
+type linkedList =
+  | Node(int, linkedList)
+  | Empty;
 ```
 
-Here, `Instagram` holds a `string`, and `Facebook` holds a `string` and an `int`. Usage:
+The constructors are now like functions that accept their data as arguments.
```

Comment:
Might be confusing as some people will try to pass them as first class functions (I _*wish*_!)

Reply:
Yeah I see, I can remove the comparison